deepspeed/runtime/zero/partition_parameters.py

```diff
@@ -1250,75 +1280,99 @@ def all_gather_coalesced(params: Iterable[Parameter],
                     return AllGatherHandle(handle, param, quantization=quant_info)
 
             else:
-                if not quantize:
-                    dtype_params = defaultdict(list)
-                    for p in params:
-                        dtype_params[p.ds_tensor.dtype].append(p)
-                    handles = []
-                    for dtype, params in dtype_params.items():
-                        handles.append(_all_gather_dtype(dtype, params, world_size, rank_in_group, ds_process_group))
+                if self.use_all_reduce_for_fetch_params and not quantize and not use_secondary_tensor:
```

Comment:
> Hi @deepcharm , these slides are cool and make sense to me. But as 2) Proposed Optimization, it showed removing unnecessay data concat&copy by avoiding params interleaving of allgather (Not allreduce). Allreduce is what confuses me, we don't do any sum/avg operation on collected weights right?

Hi @GuanhuaWang, using allreduce allows params to occupy the flat_buffer in a contiguous manner:
Each rank copies its shard and the subsequent allreduce results in all the shards restore the entire param.
The sums are between the zeros in the appropriate flat_buffer locations and the shards from a rank.
It may look a bit tricky, indeed.

Reply:
> The not quantize and not use_secondary_tensor restrict this optimization from being applied to zero++ (or hpz). I don't see a fundamental reason for this limitation, so can you please share your concerns? Or is it matter of needing more work, such as refactoring this logic to enable this optimization more generally?

@tjruwase, True, this optimization can be also applied to zero++ (or hpz).
It's just a matter of some more work and testing of those flows.